File: Top_100_Liked_Questions/208._Implement_Trie_(Prefix_Tree)/solution.py
class TrieNode:
    
    def __init__(self):
        self.links = [None] * 26
        self.end = False

    def get(self, ch):
        return self.links[self.ch2Int(ch)]
    
    def containsKey(self, ch):
        return self.get(ch) is not None
    
    def put(self, ch, node):
        self.links[self.ch2Int(ch)] = node
    
    def ch2Int(self, ch):
        return ord(ch) - ord('a')
    
    def isEnd(self):
        return self.end
    
    def setEnd(self):
        self.end = True
        
        
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        node = self.root
        for ch in word:
            if not node.containsKey(ch):
                node.put(ch, TrieNode())
            node = node.get(ch)
        
        node.setEnd()

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        node = self.root
        for ch in word:
            if not node.containsKey(ch):
                return False
            
            node = node.get(ch)
        
        return node.isEnd()

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        node = self.root
        for ch in prefix:
            if not node.containsKey(ch):
                return False
            
            node = node.get(ch)
        
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: Top_100_Liked_Questions/208._Implement_Trie_(Prefix_Tree)/solution.js
var TrieNode = function() {
    this.links = new Array(26).fill(null);
    this.end = false;
}

TrieNode.prototype.ch2Int = function(ch) {
    return ch.charCodeAt(0) - 'a'.charCodeAt(0);
}

TrieNode.prototype.containsKey = function(ch) {
    return this.get(ch) != null;
}

TrieNode.prototype.get = function(ch) {
    return this.links[this.ch2Int(ch)];
}

TrieNode.prototype.put = function(ch, node) {
    this.links[this.ch2Int(ch)] = node;
}

TrieNode.prototype.isEnd = function() {
    return this.end;
}

TrieNode.prototype.setEnd = function() {
    this.end = true;
}

/**
 * Initialize your data structure here.
 */
var Trie = function() {
    this.root = new TrieNode();
};

/**
 * Inserts a word into the trie. 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    let node = this.root;
    for (let ch of word) {
        if (!node.containsKey(ch)) {
            node.put(ch, new TrieNode());
        }
        node = node.get(ch);
    }
    node.setEnd();
};

/**
 * Returns if the word is in the trie. 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    let node = this.root;
    for (let ch of word) {
        if (!node.containsKey(ch)) {
            return false;
        }
        node = node.get(ch);
    }
    return node.isEnd();
};

/**
 * Returns if there is any word in the trie that starts with the given prefix. 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    let node = this.root;
    for (let ch of prefix) {
        if (!node.containsKey(ch)) {
            return false;
        }
        node = node.get(ch);
    }
    return true;
};

/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */